GROMACS docs (#87)
eth-cscs/cscs-docs: docs/software/sciapps/gromacs.md

[](){#ref-uenv-gromacs}
# GROMACS

[GROMACS](https://www.gromacs.org) (GROningen Machine for Chemical Simulations) is a versatile and widely-used open source package to perform molecular dynamics, i.e. simulate the Newtonian equations of motion for systems with hundreds to millions of particles.

It is primarily designed for biochemical molecules like proteins, lipids and nucleic acids that have a lot of complicated bonded interactions, but since GROMACS is extremely fast at calculating the nonbonded interactions (that usually dominate simulations) many groups are also using it for research on non-biological systems, e.g. polymers.

> 

Reviewer: Can you please add the note like in the other sciapps pages mentioning uenv and linking to the appropriate documentation?

4. **Analysis and Visualization**: GROMACS offers a suite of analysis tools to extract and analyze data from MD simulations. It provides functionalities for computing properties such as energy, temperature, pressure, radial distribution functions, and free energy landscapes. GROMACS also supports visualization tools, allowing users to visualize and analyze the trajectories of molecular systems.

5. **User-Friendly Interface**: GROMACS provides a command-line interface (CLI) and a set of well-documented input and control files, making it accessible to both novice and expert users. It offers flexibility in defining system parameters, simulation conditions, and analysis options through easily modifiable input files.

6. **Integration with Other Software**: GROMACS can be integrated with other software packages and tools to perform advanced analysis and extend its capabilities. It supports interoperability with visualization tools like VMD and PyMOL, analysis packages like GROMACS Analysis Tools (GROMACS Tools) and MDAnalysis, and scripting languages such as Python, allowing users to leverage a wide range of complementary tools.

## Licensing Terms & Conditions

> GROMACS is a joint effort, with contributions from developers around the world: users agree to acknowledge use of GROMACS in any reports or publications of results obtained with the Software (see [GROMACS Homepage](https://www.gromacs.org/about.html) for details).

Reviewer: I'd make this a note like for other applications, and move it to the end of the `# GROMACS` section (before `## Key Features`)

Author: Done

## Daint on Alps (GH200)

### Setup

> On ALPS, we provide pre-built user environments containing GROMACS alongside all the required dependencies for the GH200 hardware setup. To access the `gmx_mpi` executable, we do the following:

Reviewer 2: ```suggestion
On Alps, we provide pre-built user environments containing GROMACS alongside all the required dependencies for the GH200 hardware setup. To access the `gmx_mpi` executable, we do the following:
```
? Honestly not quite sure what the right capitalization is...

```bash
uenv image find 							  # list available images

uenv image pull gromacs/VERSION:TAG.          # copy version:tag from the list above
uenv start gromacs/VERSION:TAG --view=gromacs # load the gromacs view

gmx_mpi --version                             # check GROMACS version
```

The images also provide two alternative views, namely `plumed` and `develop`.

> After starting the pulled image using `uenv start ...` , one may do the following see the available views.

Reviewer: Maybe this is a bit redundant with the uenv documentation, and it might be better to focus on the view? (The example is also loading the `gromacs` view above, but it is introduced below.)

```bash
$ uenv status
/user-environment:gromacs-gh200
  GPU-optimised GROMACS with and without PLUMED, and the toolchain to build your own GROMACS.
  modules: no modules available
  views:
    develop
    gromacs
    plumed
```

> The `develop` view has all the required dependencies or GROMACS without the program itself. This is meant for those users who want to use a customized variant of GROMACS for their simulation which they build from source. This view makes it convenient for users as it provides the required compilers (GCC 12) along with the dependencies such as CMake, CUDA, hwloc, Cray MPICH, among many others which their GROMACS can use during build and installation. Users must enable this view each time they want to use their **custom GROMACS installation**.

Reviewer: ```suggestion
The `develop` view has all the required dependencies or GROMACS without the program itself. This is meant for those users who want to use a customized variant of GROMACS for their simulation which they build from source. This view makes it convenient for users as it provides the required compilers (GCC) along with the dependencies such as CMake, CUDA, hwloc, Cray MPICH, among many others which their GROMACS can use during build and installation. Users must enable this view each time they want to use their **custom GROMACS installation**.
```

> The `plumed` view contains GROMACS 2022.5 (older version) with PLUMED 2.9.0. This is due to the compatibility requirements of PLUMED. CSCS will periodically update these user environment images to feature newer versions as they are made available.

Reviewer: This is very specific to the current image. IIRC, the new image will have the same version of GROMACS in both views. Is there a way to make it more general to avoid going out of date?

The CUDA MPS wrapper here: [Oversubscription of GPU cards](https://confluence.cscs.ch/spaces/KB/pages/859015189/Oversubscription+of+GPU+cards)

The wrapper script above needs to be made executable with `chmod +x mps-wrapper.sh`.
 
The SLURM submission script can be adapted from the template below to use the application and the `mps-wrapper.sh` in conjunction.

```bash title="launch.sbatch"
#!/bin/bash

#SBATCH --job-name="JOB NAME"
#SBATCH --nodes=1             # number of GH200 nodes with each node having 4 CPU+GPU
#SBATCH --ntasks-per-node=8   # 8 MPI ranks per node
#SBATCH --cpus-per-task 32    # 32 OMP threads per MPI rank
#SBATCH --account=ACCOUNT
#SBATCH --hint=nomultithread  

export MPICH_GPU_SUPPORT_ENABLED=1
export FI_CXI_RX_MATCH_MODE=software

export GMX_GPU_DD_COMMS=true
export GMX_GPU_PME_PP_COMMS=true
export GMX_FORCE_UPDATE_DEFAULT_GPU=true
export GMX_ENABLE_DIRECT_GPU_COMM=1
export GMX_FORCE_GPU_AWARE_MPI=1

srun ./mps-wrapper.sh gmx_mpi mdrun -s input.tpr -ntomp 32 -bonded gpu -nb gpu -pme gpu -pin on -v -noconfout -dlb yes -nstlist 300 -gpu_id 0123 -npme 1 -nsteps 10000 -update gpu
```

> This can be run using `sbatch launch.sbatch` on the login node with the user environment loaded.

Reviewer: Instead of this, I'd add `--uenv` and `--view` to the submission script, which IMHO is a better practice.

Author: Done

	- When running on a single GPU, one can either configure the simulation with 1-2 MPI ranks with `-gpu_id` as `0`, or try running the simulation with a small number of parameters and let GROMACS run with defaults/inferred parameters with a command like the following in the SLURM script:
	`srun ./mps-wrapper.sh -- gmx_mpi mdrun -s input.tpr -ntomp 64`
	- Given the compute throughput of each Grace-Hopper module (single CPU+GPU), **for smaller-sized problems, it is possible that a single-GPU run is the fastest**. This may happen when the overheads of domain decomposition, communication and orchestration exceed the benefits of parallelism across multiple GPUs. In our test cases, a single Grace-Hopper module (1 CPU+GPU) has consistently shown a 6-8x performance speedup over a single node on Piz Daint (Intel Xeon Broadwell + P100).
	- Try runs with and without specifying the GPU IDs explicitly with `-gpu_id 0123`. For the multi-node case, removing it might yield the best performance.

## Scaling

> Benchmarking done with large MD simulation problems of 1.4 million and 3 million atom systems in order to fully saturate the GPUs from the [HECBioSim Benchmark Suite](https://www.hecbiosim.ac.uk/access-hpc/benchmarks).

Reviewer: ```suggestion
Benchmarking is done with large MD simulation of system of 1.4 million and 3 million atoms, in order to fully saturate the GPUs, from the [HECBioSim Benchmark Suite](https://www.hecbiosim.ac.uk/access-hpc/benchmarks).
```

| 4      |  30.289 | 2.11x   |

!!! warning "Known Performance/Scaling Issues"

	- The currently provided build of GROMACS allows **only one MPI rank to be dedicated for PME** with `-nmpe 1`. This becomes a serious performance limitation for larger systems where the non-PME ranks finish their work before the PME rank leading to unwanted load imbalances across ranks. This limitation is targeted to be fixed in the subsequent releases of our builds of user environments.
	- The above problem is especially critical for large problem sizes (1+ million atom systems) but is far less apparent in small and medium sized runs.
	- If the problem allows the integration step to take place on the GPU with `-update gpu`, that can lead to significant performance and scaling gains as it allows an even greater part of the computations to take place on the GPU.
	- A single node of the GH200 cluster offers 4x CPU+GPU. For problems that can benefit from scaling beyond a single node, use the flag `export FI_CXI_RX_MATCH_MODE=software` in the SBATCH script. The best use of resources in terms of node-hours might be achieved on a single node for most simulations.

## Further Documentation 

> [GROMACS Homepage](https://www.gromacs.org)
> 
> [GROMACS Manual](https://manual.gromacs.org/2024.1/index.html)

Reviewer: ```suggestion
* [GROMACS Homepage](https://www.gromacs.org)
* [GROMACS Manual](https://manual.gromacs.org/2024.1/index.html)
```
